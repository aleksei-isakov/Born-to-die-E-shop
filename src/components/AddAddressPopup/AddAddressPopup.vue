<template>
  <div
    v-if="isPopupVisible"
    class="popup-modal-wrapper"
    @click="onClickClosePopup"
  >
    <div class="popup-container" @click.stop>
      <div class="popup-container__header">
        <h1 class="popup-container__header__title">Add new address</h1>

        <v-btn color="primary" icon text @click="onClickClosePopup">
          <v-icon> close </v-icon>
        </v-btn>
      </div>

      <v-form v-model="valid" @submit.prevent="onSubmit">
        <div class="form_wrapper">
          <v-row>
            <v-col class="form-title" md="2">
              <v-select
                v-model="userdata.title"
                :items="selectTitle"
                label="Title"
                :rules="noEmptyFieldsRule"
                required
              ></v-select>
            </v-col>

            <v-col md="5">
              <v-text-field
                v-model="userdata.name"
                :rules="nameSurnameRules"
                label="Name"
                required
              ></v-text-field>
            </v-col>

            <v-col md="5">
              <v-text-field
                v-model="userdata.surname"
                :rules="nameSurnameRules"
                label="Surname"
                required
              ></v-text-field>
            </v-col>
          </v-row>

          <v-row>
            <v-col md="3">
              <v-select
                v-model="userdata.phoneNumberCode"
                :items="selectPhoneCode"
                :rules="noEmptyFieldsRule"
                label="Code"
                required
                prepend-inner-icon="phone"
              ></v-select>
            </v-col>

            <v-col md="9">
              <v-text-field
                v-model="userdata.phoneNumber"
                v-mask="'(###)###-##-######'"
                :rules="phoneRules"
                label="Phone number"
                required
              ></v-text-field>
            </v-col>
          </v-row>

          <v-row>
            <v-col md="6">
              <v-autocomplete
                v-model="userdata.country"
                :items="selectCountry"
                label="Country"
                :rules="noEmptyFieldsRule"
                persistent-hint
                required
                @change="changeCityList"
              ></v-autocomplete>
            </v-col>

            <v-col md="6">
              <v-autocomplete
                v-model="userdata.city"
                :items="selectCity"
                label="City"
                :rules="cityRules"
                persistent-hint
                required
                @input="changeCityList"
              ></v-autocomplete>
            </v-col>
          </v-row>

          <v-row>
            <v-col md="12">
              <v-text-field
                v-model="userdata.street"
                :rules="noEmptyFieldsRule"
                label="Street"
                required
              ></v-text-field>
            </v-col>
          </v-row>

          <v-row>
            <v-col md="6">
              <v-text-field
                v-model="userdata.building"
                label="Building"
                :rules="noEmptyFieldsRule"
                required
              ></v-text-field>
            </v-col>

            <v-col md="6">
              <v-text-field
                v-model="userdata.flat"
                label="Flat"
                :rules="flatRules"
                required
              ></v-text-field>
            </v-col>
          </v-row>

          <v-row>
            <v-col md="12">
              <v-text-field
                v-model="userdata.zip"
                :rules="zipRules"
                :counter="10"
                label="Zip"
                required
              ></v-text-field>
            </v-col>
          </v-row>
        </div>

        <v-card-actions>
          <base-text-filled-button
            class="form_wrapper__submit-button"
            :disabled="!valid"
            type="submit"
          >
            ADD
          </base-text-filled-button>
        </v-card-actions>
      </v-form>
    </div>
  </div>
</template>

<script>
import addressDataMock from './addressDataMock.json';
import { TITLE, PHONECODESLIST } from './helper';
import { BaseTextFilledButton } from '@/base_components';

export default {
  name: 'AddAddressPopup',

  components: { BaseTextFilledButton },

  props: {
    isPopupVisible: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    userdata: {
      title: '',
      name: '',
      surname: '',
      phoneNumberCode: '',
      phoneNumber: '',
      country: '',
      city: '',
      street: '',
      building: '',
      flat: '',
      zip: ''
    },
    selectTitle: TITLE,
    selectCountry: Object.keys(addressDataMock),
    selectPhoneCode: PHONECODESLIST,
    selectCity: null,
    isSpacesAllowed: false,
    isLettersAllowed: false,
    isEmptyFieldAllowed: false,
    maxZipLength: 10,
    minZipLength: 5,
    minNameLength: 2,
    minPhoneLength: 15,
    valid: false
  }),

  computed: {
    noEmptyFields() {
      if (!this.isEmptyFieldAllowed) {
        return (fieldValue) => !!fieldValue || 'This field is required';
      }

      return 'ok';
    },

    onlyCharacters() {
      if (!this.isLettersAllowed) {
        return (fieldValue) =>
          /^\D+$/.test(fieldValue) || 'Only characters required';
      }

      return true;
    },

    minName() {
      if (this.minNameLength) {
        return (fieldValue) =>
          (fieldValue || '').length >= this.minNameLength ||
          'A minimum of 2 characters is allowed';
      }

      return true;
    },

    minPhone() {
      if (this.userdata.phoneNumber.length < this.minPhoneLength) {
        return 'A minimum of 10 characters is allowed';
      }

      return true;
    },

    noSpaceAllowed() {
      if (!this.isSpacesAllowed) {
        return (fieldValue) =>
          (fieldValue || '').indexOf(' ') < 0 || 'No spaces allowed';
      }

      return false;
    },

    onlyNumbers() {
      if (!this.isLettersAllowed) {
        return (fieldValue) =>
          /^\d+$/.test(fieldValue) || 'Only numbers allowed';
      }

      return false;
    },

    noSpaces() {
      if (!this.isSpacesAllowed) {
        return (v) => (v || '').indexOf(' ') < 0 || 'No spaces are allowed';
      }

      return true;
    },

    selectCountryFirst() {
      if (!this.userdata.country) {
        return 'Please, select the country first';
      }

      return true;
    },

    zipCodeNumbersDash() {
      if (!this.isLettersAllowed) {
        return (fieldValue) =>
          /^\d+(-\d+)*$/.test(fieldValue) ||
          'Only numbers and a single dash is allowed';
      }

      return true;
    },

    zipCodeMaxCount() {
      if (this.userdata.zip.length > this.maxZipLength) {
        return `A maximum of ${this.maxZipLength} characters is allowed`;
      }

      return true;
    },
    zipCodeMinCount() {
      if (this.userdata.zip.length < this.minZipLength) {
        return `A minimum of ${this.minZipLength} characters is allowed`;
      }

      return true;
    },

    noEmptyFieldsRule() {
      return [this.noEmptyFields];
    },

    nameSurnameRules() {
      return [this.noEmptyFields, this.onlyCharacters, this.minName];
    },

    phoneRules() {
      return [this.noEmptyFields, this.minPhone];
    },

    flatRules() {
      return [this.noEmptyFields, this.noSpaces, this.onlyNumbers];
    },

    cityRules() {
      return [this.selectCountryFirst, this.noEmptyFields];
    },

    zipRules() {
      return [
        this.noEmptyFields,
        this.zipCodeMaxCount,
        this.zipCodeNumbersDash,
        this.zipCodeMinCount
      ];
    }
  },

  methods: {
    onSubmit() {},

    onClickTogglePopup() {
      this.props.isPopupVisible = !this.props.isPopupVisible;
    },

    onClickClosePopup() {
      this.$emit('onClickClosePopup');
    },
    changeCityList() {
      this.selectCity = addressDataMock[this.userdata.country];
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.popup-modal-wrapper {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: $z-index-popup;
  display: flex;
  background-color: rgba(0, 0, 0, 0.5);
}

.popup-container {
  position: relative;
  width: 600px;
  margin: auto;
  padding: 20px;
  background-color: $white;
  border-radius: 5px;

  &__header {
    display: flex;
    justify-content: space-between;
    padding-left: 20px;

    &__title {
      font-weight: normal;
    }
  }
}

.form_wrapper {
  width: 100%;
  padding: 20px 20px;

  &__submit-button {
    width: 100%;
  }
}

@media (max-width: $tablet-size) {
  .popup-container {
    width: 400px;

    &__header {
      justify-content: right;

      &__title {
        display: none;
      }
    }
  }

  .form_wrapper {
    width: 100%;
    padding: 20px 20px;
  }

  .form-title {
    flex-basis: 100%;
  }
}
</style>
