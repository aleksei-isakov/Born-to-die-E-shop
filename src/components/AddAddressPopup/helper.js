export const PHONECODESLIST = [
  '+7',
  '+44',
  '+1',
  '+90',
  '+34',
  '+39',
  '+49',
  '+33'
];
export const TITLE = ['Mr.', 'Mrs.', 'Ms.'];
