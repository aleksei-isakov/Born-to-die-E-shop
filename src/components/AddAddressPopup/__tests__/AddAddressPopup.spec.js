import Vue from 'vue';
import { mount, createLocalVue } from '@vue/test-utils';
import AddAddressPopup from '@/components/AddAddressPopup/AddAddressPopup';
import Vuetify from 'vuetify';
Vue.directive('mask', {});

let wrapper;
const userdata = {
  title: 'Mr',
  name: 'Arkadiy',
  surname: 'Dobkin',
  phoneNumberCode: '+44',
  phoneNumber: '(999)6659986',
  country: 'Russia',
  city: 'Moscow',
  street: 'Street',
  building: '1',
  flat: '1',
  zip: '55555-5555'
};
const isPopupVisible = true;

Vue.use(Vuetify);
const vuetify = new Vuetify();

describe('AddAddressPopup', () => {
  afterEach(() => {
    wrapper.destroy();
  });

  it('submit button should be disabled if form fields are empty', async () => {
    const localVue = createLocalVue();

    wrapper = mount(AddAddressPopup, {
      localVue,
      vuetify,
      propsData: {
        isPopupVisible: isPopupVisible
      },
      data() {
        return {
          userdata: {
            title: '',
            name: '',
            surname: '',
            phoneNumberCode: '',
            phoneNumber: '',
            country: '',
            city: '',
            street: '',
            building: '',
            flat: '',
            zip: ''
          }
        };
      }
    });

    await Vue.nextTick();

    expect(wrapper.vm.$data.valid).toBeFalsy();
    expect(wrapper.find('.form_wrapper__submit-button').exists()).toBeTruthy();
    expect(
      wrapper.find('.form_wrapper__submit-button').element.disabled
    ).toBeTruthy();
  });

  it('submit button should be active if all fields are filled out properly', async () => {
    const localVue = createLocalVue();

    wrapper = mount(AddAddressPopup, {
      localVue,
      vuetify,
      propsData: {
        isPopupVisible: isPopupVisible
      },
      data() {
        return {
          userdata
        };
      }
    });

    await Vue.nextTick();

    expect(wrapper.vm.$data.valid).toBeTruthy();
    expect(wrapper.find('.form_wrapper__submit-button').exists()).toBeTruthy();
    expect(
      wrapper.find('.form_wrapper__submit-button').element.disabled
    ).toBeFalsy();
  });

  it('should render correctly and match the snapshot', () => {
    const localVue = createLocalVue();

    wrapper = mount(AddAddressPopup, {
      localVue,
      vuetify,
      propsData: {
        isPopupVisible: isPopupVisible
      },
      data() {
        return {
          userdata: {
            title: '',
            name: '',
            surname: '',
            phoneNumberCode: '',
            phoneNumber: '',
            country: '',
            city: '',
            street: '',
            building: '',
            flat: '',
            zip: ''
          }
        };
      }
    });
    expect(wrapper).toMatchSnapshot();
  });
});
