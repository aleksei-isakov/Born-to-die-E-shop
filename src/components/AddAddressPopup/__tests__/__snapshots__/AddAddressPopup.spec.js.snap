// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAddressPopup submit button should be disabled if form fields are empty 1`] = `
<div class="popup-modal-wrapper">
  <div class="popup-container">
    <div class="popup-container__header">
      <h1 class="popup-container__header__title">Add new address</h1> <button type="button" class="v-btn v-btn--icon v-btn--round v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">close</i></span></button>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="form_wrapper">
        <div class="row">
          <div class="form-title col-md-2 col">
            <div class="v-input theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-5" class="v-input__slot">
                  <div class="v-select__slot"><label for="input-5" class="v-label theme--light" style="left: 0px; position: absolute;">Title</label>
                    <div class="v-select__selections"><input required="required" id="input-5" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-md-5 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Name</label><input required="required" id="input-10" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-md-5 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-13" class="v-label theme--light" style="left: 0px; position: absolute;">Surname</label><input required="required" id="input-13" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md-3 col">
            <div class="v-input theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-16" class="v-input__slot">
                  <div class="v-input__prepend-inner">
                    <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">phone</i></div>
                  </div>
                  <div class="v-select__slot"><label for="input-16" class="v-label theme--light" style="left: 0px; position: absolute;">Code</label>
                    <div class="v-select__selections"><input required="required" id="input-16" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-md-9 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-22" class="v-label theme--light" style="left: 0px; position: absolute;">Phone number</label><input required="required" id="input-22" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md-6 col">
            <div class="v-input theme--light v-text-field v-select v-autocomplete">
              <div class="v-input__control">
                <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-25" class="v-input__slot">
                  <div class="v-select__slot"><label for="input-25" class="v-label theme--light" style="left: 0px; position: absolute;">Country</label><input required="required" id="input-25" type="text" autocomplete="off">
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-md-6 col">
            <div class="v-input theme--light v-text-field v-select v-autocomplete">
              <div class="v-input__control">
                <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-30" class="v-input__slot">
                  <div class="v-select__slot"><label for="input-30" class="v-label theme--light" style="left: 0px; position: absolute;">City</label><input required="required" id="input-30" type="text" autocomplete="off">
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                    </div><input type="hidden">
                  </div>
                  <div class="v-menu">
                    <!---->
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md-12 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-35" class="v-label theme--light" style="left: 0px; position: absolute;">Street</label><input required="required" id="input-35" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md-6 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-38" class="v-label theme--light" style="left: 0px; position: absolute;">Building</label><input required="required" id="input-38" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-md-6 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-41" class="v-label theme--light" style="left: 0px; position: absolute;">Flat</label><input required="required" id="input-41" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md-12 col">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-44" class="v-label theme--light" style="left: 0px; position: absolute;">Zip</label><input required="required" id="input-44" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                  <div class="v-counter theme--light">0 / 10</div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions"><button class="base-button base-text-button base-text-filled-button form_wrapper__submit-button custom-blue" type="submit" disabled="disabled">
          ADD
        </button></div>
    </form> <button>show</button>
  </div>
</div>
`;
